src/main/scala/lantern/dslapi.scala

```diff
@@ -206,6 +169,21 @@ trait DslExp extends Dsl
   }
 }
 
+trait TensorOps extends DslOps {
+  object NewGPUArray {
+    // Allocate an array of the specified size on the GPU.
+    def apply[T: Manifest](scalarCount: Rep[Int]): Rep[Array[T]] = gpu_array_new(scalarCount)
+  }
+  def gpu_array_new[T: Manifest](scalarCount: Rep[Int]): Rep[Array[T]] = ???
```

Comment:
Usually the XXXOps file just declare the function, don't need the ???. Otherwise it will turn a compiler error (no implementation found) into a runtime error, UnimplementedException

Reply:
Without `???`, all the tests that extend `DslDriverC` with `TensorOps` need to define `gpu_array_new`, even if they don't use GPU ops at all. This is unergonomic.

Example:
```scala
val mnist2  = new DslDriverC[String, Unit] with NNModule with ScannerLowerExp {
  // `NNModule` extends `TensorDsl`, which extends `TensorOps`.
  // `gpu_array_new ` is unimplemented.
```

One alternative is to merge `DslOps/DslExp` and `TensorOps/TensorExp` for simplicity. Code generators (e.g. `DslGenScala` and `DslGenBase`) are then responsible for handling `GPUArrayNew`, and they can use `???`.

What are your thoughts?

Reply:
I would create a separate trait GPUAllocOps and mix it only for Cuda code not into the LanternC. But create a LanternCuda for example.

But this nit picking I guess.

Reply:
I see.

The problem is `NewGPUArray` is called within `TensorDsl`, which `LanternDriverC` mixes in (by way of `LanternDriverC` <- `LanternDriver <- TensorDslExp <- TensorDsl`).

We could make a `TensorDslCuda` trait that extends `TensorDsl` and mixes in "GPUAllocOps". `BackendCublas`/`BackendCudnn` would then be defined in `TensorDslCuda`. I can do this sometime.